overlord/devicestate/devicestate.go

```diff
@@ -461,6 +467,15 @@ func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSna
 			changed = ms.currentModelSnap.PinnedTrack != ms.newModelSnap.PinnedTrack
 		}
 		if changed {
+			logger.Debugf("essential snap %s track changed", ms.newModelSnap.Name)
+			if flags.localSnapsRequired {
+				if localSi == nil {
+					return nil, fmt.Errorf("no file provided for %q snap (track changed)", ms.newModelSnap.SnapName())
+				}
+				return snapstateUpdatePathWithDeviceContext(st, ms.newSnap,
+					&snapstate.RevisionOptions{Channel: newModelSnapChannel},
+					userID, snapstate.Flags{NoReRefresh: true}, deviceCtx, fromChange, localSi, path)
+			}
 			// new modes specifies the same snap, but with a new channel
 			return snapstateUpdateWithDeviceContext(st, ms.newSnap,
```

Comment:
maybe all these "if"s on localSnapsRequired can be expressed as methods on the remodelFlags instead